modyn/supervisor/internal/triggers/datadrifttrigger.py

```diff
@@ -127,41 +154,85 @@ def inform(
-            # If we've never triggered before, always trigger
-            self._triggered_once = True
-            triggered = True
-            drift_results: dict[str, MetricResult] = {}
-        else:
-            # Run the detection
-            triggered, drift_results = self._run_detection()
-
-        yield from self._handle_drift_result(triggered, trigger_idx, drift_results, log=log)
+
+        # index of the first unprocessed data point in the batch
+        processing_head_in_batch = 0
 
         # Go through remaining data in new data in batches of `detect_interval`
-        for i in range(0, len(new_key_ts), detect_interval):
-            batch = new_key_ts[i : i + detect_interval]
-            trigger_idx += detect_interval
-            self._update_curr_window(batch)
-
-            if len(batch) == detect_interval:
-                # Regular batch, in this case run detection
-                triggered, drift_results = self._run_detection()
+        while True:
+            if self._sample_left_until_detection - len(new_key_ts) > 0:
+                # No detection in this trigger because of too few data points to fill detection interval
+                self._windows.inform_data(new_key_ts)  # update current window
+                self._sample_left_until_detection -= len(new_key_ts)
+                return
+
+            # At least one detection, fill up window up to that detection
+            next_detection_interval = new_key_ts[: self._sample_left_until_detection]
+            self._windows.inform_data(next_detection_interval)
+
+            # Update the remaining data
+            processing_head_in_batch += len(next_detection_interval)
+            new_key_ts = new_key_ts[len(next_detection_interval) :]
+
+            # Reset for next detection
+            self._sample_left_until_detection = self.config.detection_interval_data_points
+
+            # The first ever detection will always trigger
+            if not self._triggered_once or len(self.warmup_intervals) < (self.config.warmup_intervals or 0):
+                # If we've never triggered before, always trigger
+                self._triggered_once = True
+                triggered = True
+                drift_results: dict[str, MetricResult] = {}
+                if len(self.warmup_intervals) < (self.config.warmup_intervals or 0):
+                    self.warmup_intervals.append(list(self._windows.reference))
+
+            else:
+                # Run the detection
+
+                # if this is the first non warmup detection, we inform the metrics that use decision criteria
+                # with calibration requirements about the warmup intervals so they can calibrate their thresholds
+                if len(self.warmup_intervals) > 0:
+                    # we can ignore the results as the decision criteria will keep track of the warmup results
+                    # internally
+                    if self._any_metric_needs_calibration():
```

Comment:
According to codecov, the calibration/warmup logic does not seem to be tested here

Reply:
Yes, I will add a test for the warmup logic